nixos/radicle: init services

ju1m opened the pull request:
## Description of changes

- [X] Add `radicle-node.service`
- [X] Add `radicle-httpd.service`
  - [X] ~Add `radicle-httpd` to `pkgs.radicle-node`~ `radicle-httpd` has been put in a separate package in another PR.
- [X] Add `nixosTests.radicle`
  - [X] Add test for `radicle-httpd`

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [X] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [X] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

ju1m committed:
nixos/radicle: init services

gador commented on nixos/tests/radicle.nix:
For this test to run, there is `radicle = handleTest ./radicle.nix {};` missing from `all-tests.nix`

ju1m replied:
Oh, thanks, I missed that because I was using `nix -L build -f nixos/tests/radicle.nix`, now ofborg should be able to see the test.

gador commented on nixos/modules/services/misc/radicle.nix:
```diff
@@ -0,0 +1,515 @@
+{ config, lib, pkgs, options, ...  }:
```
```suggestion
{ config, lib, pkgs, ...  }:
```

ju1m replied:
`options` is actually used to fetch the type of an nginx virtualHost.

It's like what's done in some modules in `nixos/modules/services/web-apps/`:
```
type = with types; nullOr (submodule (import ../web-servers/nginx/vhost-options.nix { inherit config lib; }));
```
but without re-importing any file.

gador commented on nixos/modules/services/misc/radicle.nix:
```diff
@@ -0,0 +1,515 @@
+                type = types.ints.positive;
+                default = 7 * 24 * 60 * 60; # 1 week
+                defaultText = "7 * 24 * 60 * 60";
+                description = "How long to keep a routing table entry before being pruned";
```
```suggestion
                description = "How many seconds to keep a routing table entry before being pruned.";
```

ju1m replied:
I copied verbatim the description from https://app.radicle.xyz/nodes/seed.radicle.garden/rad:z3gqcJUoA1n9HaHKufZs5FCSGazv5/tree/radicle/src/node/config.rs#L99 
Let's make an exception for clarity ;-)

gador commented on nixos/modules/services/misc/radicle.nix:
```diff
@@ -0,0 +1,515 @@
+      commonConfig = serviceName:
+        {
+          environment = env // {
+            RUST_BACKTRACE = "1";
+            RUST_LOG = mkDefault "info";
```
is this necessary?

ju1m replied:
No, I don't think it's necessary, I took it from https://app.radicle.xyz/nodes/seed.radicle.garden/rad:z3gqcJUoA1n9HaHKufZs5FCSGazv5/tree/systemd/radicle-node.service#L24 and left it in case some hardening options were causing troubles, but I guess it does not belong to a production environment.
I'll remove [RUST_BACKTRACE](https://doc.rust-lang.org/std/backtrace/), I can always set it on my personal config if need be.

gador reviewed (commented):
Thanks for all your work towards radicle in NixOS! I like how you structured the module. I only have a few remarks.
When adding the test to `all-tests.nix` the nixosTests pass without issues :+1:

lorenzleutgeb commented:
@ju1m thanks a lot for picking this up. I am actually in Berlin with the Radicle team this week, which is why I couldn't review this today. Hopefully tomorrow! :)

lorenzleutgeb commented:
How about splitting into more packages? The web stuff (which is not necessary to run and administer a node). I have sketched that here: https://github.com/NixOS/nixpkgs/pull/309050#issuecomment-2094888831

lorenzleutgeb commented:
How about splitting into more packages? The web stuff (which is not necessary to run and administer a node). I have sketched that here: https://github.com/NixOS/nixpkgs/pull/309050#issuecomment-2094888831

ju1m commented:
> @ju1m thanks a lot for picking this up. I am actually in Berlin with the Radicle team this week, which is why I couldn't review this today. Hopefully tomorrow! :)

No hurry, thanks to the team for making such a promising tool, and thanks to you for pushing its integration into Nixpkgs, that's what caught my attention over it.
I'm not sure I'll commit to using radicle yet due to the "high" level of network I/O and CPU cycles it seems to require to run in sync with the public peers. But the tool is young, so maybe it will get lightweight enough for me at some point. In any case I'm adding myself as a maintainer for now to keep being notified if some help is needed in the future for that service.

lorenzleutgeb commented:
> Sure, though not being used to Rust packaging I went to the most expedient in this PR. I think the splitting can go in another PR, I'll amend this PR to introduce `services.radicle.httpd.{enable,package}`.

There even is talk to move `radicle-httpd` (and I guess also `rad-web`) into a separate repo. When that happens, at the latest, the justification for the split would be more obvious. I wanted to anticipate this, and keep the changes to Nix users at a minimum (fomr `radicle-cli` to `radicle-node`, and then to  `radicle-web` is a bit noisy).

Then there's also the question whether it should be `services.radicle-{httpd,node}.*` or `services.radicle.{httpd,node}.*`. I don't know whether there is a clear policy. For example, there is `services.buildbot-{master,worker}` but also `services.salt.{master,minion}`.

> @lorenzleutgeb, no hurry, thanks to the team for making such a promising tool, and thanks to you for pushing its integration into Nixpkgs, that's what caught my attention over it. I'm not sure I'll commit to using radicle yet due to the "high" level of network I/O and CPU cycles it seems to require to run in sync with the public peers. But the tool is young, so maybe it will get lightweight enough for me at some point. In any case I'm adding myself as a maintainer for now to keep being notified if some help is needed in the future for that service.

Sweet! Yeah, there was some breakage in the protocol recently, and a nodes are now announcing the version they are running, to make upgrades smoother.

Note that high I/O and CPU is being worked on:
 - https://radicle.zulipchat.com/#narrow/stream/369873-support/topic/high.20CPU.20use.20while.20disconnected.20from.20network
 - https://app.radicle.xyz/nodes/seed.radicle.xyz/rad:z3gqcJUoA1n9HaHKufZs5FCSGazv5/patches/c354a9e6792bf3aa42c42e6ba7ee628b04c802ae

lorenzleutgeb commented on nixos/tests/radicle.nix:
```diff
@@ -0,0 +1,187 @@
+      services.radicle = {
+        enable = true;
+        privateKeyFile = pkgs.writeText "id_ed25519" ''
+          -----BEGIN OPENSSH PRIVATE KEY-----
+          b3BlbnNzaC1rZXktdjEAAAAABG5vbmUAAAAEbm9uZQAAAAAAAAABAAAAMwAAAAtzc2gtZW
+          QyNTUxOQAAACCk95gwwCfu2ezgoPOHsATixXUqvm7k2mfNx6t3lBf71gAAAJDAzemhwM3p
+          oQAAAAtzc2gtZWQyNTUxOQAAACCk95gwwCfu2ezgoPOHsATixXUqvm7k2mfNx6t3lBf71g
+          AAAECMhq3OzhskoKMnQ6nMExCcRIxchCHX75shzdy0qfJcUaT3mDDAJ+7Z7OCg84ewBOLF
+          dSq+buTaZ83Hq3eUF/vWAAAAB3JhZGljbGUBAgMEBQY=
+          -----END OPENSSH PRIVATE KEY-----
```
Instead of checking in a keypair to Nixpkgs, maybe reuse one?

 - https://github.com/NixOS/nixpkgs/blob/f478e1cb40ff17ef02ab22f45e0512868b0c86c6/nixos/tests/ssh-keys.nix
 - https://github.com/NixOS/nixpkgs/tree/f478e1cb40ff17ef02ab22f45e0512868b0c86c6/nixos/modules/profiles/keys

ju1m replied:
you're right, I hesitated to use `ssh-keys.nix` for no good reason.

lorenzleutgeb commented on nixos/tests/radicle.nix:
```diff
@@ -0,0 +1,187 @@
+          node = {
+            peers = {
+              type = "dynamic";
+              # Remove noisy messages "Not enough available peers to connect to"
+              # in driverInteractive
+              target = 0;
```
If you really just want this to silence the interactive driver, then set on [`interactive.nodes.need.services.radicle.settings.peers.target = 0`](https://nixos.org/manual/nixos/stable/#sec-nixos-test-interactive-configuration)?

lorenzleutgeb commented on nixos/modules/services/misc/radicle.nix:
```diff
@@ -0,0 +1,523 @@
+            Set to {} if you do not need any customization to the virtual host.
+            If enabled, then by default, the {option}`serverName` is
+            `radicle-''${config.networking.hostName}.''${config.networking.domain}`,
+            SSL is active, and certificates are acquired via ACME.
```
```suggestion
            TLS is active, and certificates are acquired via ACME.
```

Maybe... It's a nit..

ju1m replied:
Ah right, that's what happens when one blindly copies/pastes from other modules :D

lorenzleutgeb commented on nixos/modules/services/misc/radicle.nix:
```diff
@@ -0,0 +1,523 @@
+          '';
+          description = ''
+            With this option, you can customize an nginx virtual host which already has sensible defaults for `radicle-httpd`.
+            Set to {} if you do not need any customization to the virtual host.
```
```suggestion
            Set to `{}` if you do not need any customization to the virtual host.
```

lorenzleutgeb commented on nixos/modules/services/misc/radicle.nix:
```diff
@@ -0,0 +1,523 @@
+            If enabled, then by default, the {option}`serverName` is
+            `radicle-''${config.networking.hostName}.''${config.networking.domain}`,
+            SSL is active, and certificates are acquired via ACME.
+            If this is set to null (the default), no nginx virtualHost will be configured.
```
```suggestion
            If this is set to `null` (the default), no nginx virtual host will be configured.
```

lorenzleutgeb commented on nixos/modules/services/misc/radicle.nix:
```diff
@@ -0,0 +1,523 @@
+      node = {
+        listenAddress = mkOption {
+          type = types.str;
+          default = "0.0.0.0";
```
```suggestion
          default = "0.0.0.0";
          example = "127.0.0.1";
```

lorenzleutgeb commented:
In my module, I implemented a check to catch invalid configuration in `settings`. Any particular reason not to have this?

https://github.com/radicle-nix/radicle-nix/blob/9c1adbb7d85d12df6236e006c12ac089cb18341b/os/module/services/radicle.nix#L34-L48

ju1m commented:
> In my module, I implemented a check to catch invalid configuration in `settings`. Any particular reason not to have this?

Ah thanks, I just didn't understand that it was a check. I've added a similar check.

ju1m commented:
@ofborg test radicle

spacefrogg commented:
Hi, thank you for all the work towards a radicle module.

I have one remark regarding the `settings` implementation. I would advise against creating options for each configuration value. This has shown in the past to have several drawbacks: high maintenance cost, excessively long configuration.nix man page, little informational value, migration issues. It would be best to leave the documentation of individual settings to the package's documentation (and even help create that instead of replicating it, here) and just implement the necessary bits to reliably set up the service on NixOS. This is precisely, why the settings option is just an attribute set to be converted into JSON. `checkConfig` is the better way to deal with the issue.

lorenzleutgeb commented:
Just noting: This conflicts with https://github.com/NixOS/nixpkgs/pull/313740, but I am happy to rebase my PR once this one has landed.

ju1m commented:
- [X] Remove options for each setting value according to https://github.com/NixOS/nixpkgs/pull/314440#issuecomment-2145133671
- [X] Add `pkgs.radicle-node.tests.nixos-build.{checkConfig-success,checkConfig-failure}` to test `services.radicle.checkConfig`.

ju1m commented:
> Just noting: This conflicts with #313740 (which was opened about 3 days earlier), but I am happy to rebase my PR once this one has landed.

@lorenzleutgeb, sorry for the conflict. Your PR should be easier to get merged than this bigger one so don't hold back, I can easily rebase this one.

spacefrogg commented:
The `externalAddress` setting is currently not automatically set and merged into the config, when activating the radicle-httpd service (as can be seen in the test). This will break the service for unsuspecting users. You could add the `externalAddress` option again to `settings` and make it read-only, such that it is always consistent with the module options and still inspectable via `settings`.

ju1m commented:
@spacefrogg, thanks, I wrongly thought that the NID (unknown at build time) was necessary in the `node.externalAddresses`. This setting is now initialized with a `mkDefault` when `services.radicle.httpd.enable == true` and `services.radicle.httpd.nginx != null`:
```nix
services.radicle.settings.node.externalAddresses = mkDefault [
  "${cfg.httpd.nginx.serverName}:${toString cfg.node.listenPort}"
];
```
I've not made `node.externalAddresses` read-only though because I don't understand the need for that.

The change from the previous iteration is big only because I've also grouped the `mkIf` guards to structure the module better, which helped me to spot that `services.radicle.httpd.enable` was not honored previously.

ju1m commented:
- [X] Rebase after the merge of https://github.com/NixOS/nixpkgs/pull/313740

gador commented:
Radicle v1.0.0-rc.11 [has been released](https://radicle.zulipchat.com/#narrow/stream/409174-announcements/topic/1.2E0.2E0-rc.2E11).
This includes some breaking changes, notable:
>radicle-httpd and rad-web are no longer part of this release, they are
>now part of the radicle-explorer repo and will be released separately.

Do you want to include the newer release here in the PR and update the raddicle-httpd service, or should we merge this one and update radicle alongside this module in a seperate PR?

lorenzleutgeb commented:
> should we merge this one and update radicle alongside this module in a seperate PR?

Yes. See See https://github.com/NixOS/nixpkgs/pull/321299

gador commented:
One thing I noticed is that I had to set `services.radicle.httpd.nginx = mkForce null;` Otherwise I do get 
```
error: The option `services.radicle.httpd.nginx` is defined both null and not null in [... this config file] and /nix/store/n20947dgf9ff0qp8rv99756qr3ii411y-source/nixos/modules/services/misc/radicle.nix
```
when I don't use `mkForce`.
I haven't found the reason, yet, because the default should be `null` from the module in this pr.

For reference, this is my config:
``` nix
services.radicle = {
        enable = true;
        # encrypted with systemd-creds on hel
        privateKeyFile = "radicle:${./radicle_hel.key}";
        publicKeyFile = ./hel.pub;
        settings = {
          node.alias = "gador";
          node.openFirewall = true;
        };
        httpd = {
          nginx = mkForce null;
          enable = true;
          listenAddress = "127.0.0.1";
          listenPort = 8445;
        };
      };
```

nginx is running and configured for a bunch of different vhosts on this server.

ju1m commented:
- [X] Fix `services.radicle.httpd.nginx = null;`
- [X] Rebase after https://github.com/NixOS/nixpkgs/pull/321299
- [X] Include https://github.com/NixOS/nixpkgs/pull/321580 to still have `radicle-httpd` (will drop it once it's merged).

SuperSandro2000 commented on pkgs/by-name/ra/radicle-node/package.nix:
```diff
@@ -86,6 +88,25 @@
         touch $out
       '';
+      nixos-build = lib.recurseIntoAttrs {
+        checkConfig-success =
+          let host = nixos {
+              services.radicle.settings = {
+                node.alias = "foo";
+              };
+            };
+          in
+          host.config.services.radicle.configFile;
```
```suggestion
        checkConfig-success = (nixos {
          services.radicle.settings = {
            node.alias = "foo";
          };
        }).config.services.radicle.configFile;
```

ju1m replied:
- [X] Remove `host` intermediate binding.

SuperSandro2000 commented on pkgs/by-name/ra/radicle-node/package.nix:
```diff
@@ -86,6 +88,25 @@
+            };
+          in
+          host.config.services.radicle.configFile;
+        checkConfig-failure =
+          let host = nixos {
+              services.radicle.settings = {
+                node.alias = null;
+              };
+            };
+          in
+          testers.testBuildFailure host.config.services.radicle.configFile;
```
```suggestion
        checkConfig-failure = testers.testBuildFailure (nixos {
          services.radicle.settings = {
            node.alias = null;
          };
        }).config.services.radicle.configFile;
```

ju1m replied:
- [X] Remove `host` intermediate binding.

SuperSandro2000 commented on pkgs/by-name/ra/radicle-httpd/package.nix:
```diff
@@ -0,0 +1,78 @@
+  '';
+
+  meta = {
+    description = "Radicle httpd";
+    longDescription = ''
+      A Radicle HTTP daemon exposing a JSON HTTP API that allows someone to browse local
+      repositories on a Radicle node via their web browser.
+    '';
```
```suggestion
    description = "Radicle HTTP daemon exposing a JSON HTTP API";
    longDescription = ''
      A Radicle HTTP daemon exposing a JSON HTTP API that allows someone to browse local
      repositories on a Radicle node via their web browser.
    '';
```

ju1m replied:
- [X] Drop this file, merged in the PR it was copied from.

SuperSandro2000 commented on nixos/modules/services/misc/radicle.nix:
```diff
@@ -0,0 +1,344 @@
+        ];
+        KillMode = "process";
+        StateDirectory = [ "radicle" ];
+        DynamicUser = true;
```
We really want to chown all the git file on each start?

SuperSandro2000 commented on nixos/modules/services/misc/radicle.nix:
```diff
@@ -0,0 +1,344 @@
+        KillMode = "process";
+        StateDirectory = [ "radicle" ];
+        DynamicUser = true;
+        # The "radicale" user will be allocated when the first of the "radicle-*" services starts,
+        # it will be shared among them, and be released when the last one of them exits.
```
Especially when multiple services share the same user, using DynamicUser can cause some strange behaviors.

ju1m replied:
@SuperSandro2000, what strange behavior do you have in mind? because it's [a setup suggested by Lennart Poettering](https://github.com/systemd/systemd/issues/7253#issuecomment-344289925):

> note that you can actually run multiple independent services as the same dynamic user. Wouldn't that already get you quite far? i.e. let's say you have five services that shall be able to mess around with each other files. Simply set DynamicUser=1 for all of them, but combine this with User= set to some common, clear user name. If you do, then the user will be allocated when the first of these services starts, and be released when the last one of them exits.

SuperSandro2000 commented on nixos/modules/services/misc/radicle.nix:
```diff
@@ -0,0 +1,344 @@
+        allowedTCPPorts = [ cfg.node.listenPort ];
+      };
+
+      networking.nftables.ruleset = mkIf config.networking.nftables.enable (mkBefore ''
```
Is nftables optional? If not we should assert that it is activated.

Also I don't think we need to make this here conditional. If nftables is not active, the option should just do nothing.

ju1m replied:
@SuperSandro2000, yes, `nftables` is optional, hence `config.networking.nftables.enable` is not set.
I agree that `mkIf config.networking.nftables.enable` is not needed, I'll remove it.

ju1m commented:
- [X] Rebase after the merge of https://github.com/NixOS/nixpkgs/pull/321580 (I've not re-run the checks yet, it will take the night on my computer).

ju1m commented:
- [X] Drop `DynamicUser=` for the sake of clarity, since the same level of `systemd-analyze security` can be achieved without it. Note that therefore the state directory is no longer `/var/lib/private/radicle` but `/var/lib/radicle`.

ju1m commented:
- [X] Rebase after conflict in `nixos/doc/manual/release-notes/rl-2411.section.md` following the merge of 28512d1ba4307

lorenzleutgeb commented:
I'd really like to see this merged. Any ideas on how we can get someone with write permissions on board? @SuperSandro2000 could you do another of review or merge please? 😃

dpc commented:
I just spent a lot of time debugging what came down to:

```
   services.radicle = {
     enable = true;
     privateKeyFile = "/run/secrets/radicle/seednode";
-    publicKeyFile = "/run/secrets/radicle/seednode-pub";
+    # publicKeyFile = "/run/secrets/radicle/seednode-pub";
+    publicKeyFile = "ssh-ed25519 AAAAC3NzaC1lZDI1NTE5AAAAINIxyouLgGi/RGEeBycs47NuntkAWn6ZAynImFVYsGN6 dpc@ren";
     httpd.enable = true;
     httpd.nginx.serverName = "radicle.dpc.pw";
   };
```

It seems like a gotcha or a bug that `privateKeyFile` and `publicKeyFile` are named the same way, but have different semantics.

ju1m commented:
@dpc, I'm sorry this bug made you loose your time.
Both `path` and `str` types should have worked because `services.radicle` currently uses:
```
"${if isPath cfg.publicKeyFile then cfg.publicKeyFile else pkgs.writeText "radicle.pub" cfg.publicKeyFile}:${env.RAD_HOME}/keys/radicle.pub"
```
But using `isPath` is wrong here, because it does not consider `"/run/secrets/radicle/seednode-pub"` as a path:
```console
$ nix repl
nix-repl> builtins.isPath "/run/secrets/radicle/seednode-pub"
false
nix-repl> builtins.isPath /run/secrets/radicle/seednode-pub  
true
```

Instead, `types.path.check` should have been used to also consider as a path any string prefixed by "/":
```
check = x: isStringLike x && builtins.substring 0 1 (toString x) == "/";
```

See https://github.com/NixOS/nixpkgs/pull/329731 for a fix.

dpc commented:
I'm happy to help improve it. :)

BTW. Is there any existing Wiki page/guide? After getting [my node set up](https://app.radicle.xyz/nodes/radicle.dpc.pw/), I have no idea how to do things for me.

ju1m commented:
@dpc, you can [`rad-system seed $rid`](https://radicle.xyz/guides/seeder#setting-a-repositorys-seeding-policy),
where `rad-system` is installed by `services.radicle` to execute `rad` inside the namespaces of `radicle-node.service`,
and `$rid` is given by `rad .` in your git repository.

dpc commented on nixos/modules/services/misc/radicle.nix:
```diff
@@ -0,0 +1,347 @@
+      node = {
+        listenAddress = mkOption {
+          type = types.str;
+          default = "0.0.0.0";
```
I don't know if this or the app itself is causing the app not not bind to ipv6 port.

ju1m replied:
@dpc, right, you're hitting the same problem as the test suite.
Fixed in https://github.com/NixOS/nixpkgs/pull/330021

dpc commented:
Had connectivity issues because the app wouldn't bind to ipv6.

Setting:

```
+    node.listenAddress = "[::0]";
```

fixed it.

dpc commented:
BTW. It would be great if:

```
+    settings = {
+        "web" = {
+          "pinned" = {
+            "repositories" = [
+              "rad:z2eeB9LF8fDNJQaEcvAWxQmU7h2PG"
+            ];
+          };
+        };
+    };
```

could just have its own key.

ju1m commented:
> could just have its own key. What do you think?

@dpc, you mean to add an alias to `settings.web.pinned.repositories`? what problem would it solve?
Note that [we previously reverted from type-checking `settings`](https://github.com/NixOS/nixpkgs/pull/314440#issuecomment-2145133671).

dpc commented:
@ju1m I wish i could just:

```
  services.radicle = {
    enable = true;
    # ..
    pinned = [ "rad:z2eeB9LF8fDNJQaEcvAWxQmU7h2PG" ];
  };
```

as it seems rather common need, and having to figure out and mimick config-file structure
seems like an unnecessary chore. :D

ju1m commented:
@dpc, you're raising a discoverability concern on NixOS but I will argue that it belongs to Radicle. Using `settings` and its config-mimicking, instead of inventing ad-hoc options like NixOS used to do before the `settings` framework existed, is what enables NixOS to let the burden of discoverability/maintainability onto Radicle's documentation/code where it's best addressed. Yet, in NixOS, we can address the common need you raised to set pinned repositories by setting an `example` for `settings`.
See https://github.com/NixOS/nixpkgs/pull/330021/files#diff-47ed1acddaad94538b9ee7995ffa8d7cc1376f9667350acee9cec912cec6a3bfR183

dpc commented:
Thanks! Fine with me.
